backend/grant/email/send.py

```diff
@@ -360,24 +367,47 @@ def send_email(to, type, email_args):
         sub = info['subscription']
         if user and not is_subscribed(user.settings.email_subscriptions, sub):
             print(f'Ignoring send_email to {to} of type {type} because user is unsubscribed.')
-            return
+            return None
 
-    try:
-        email = generate_email(type, email_args, user)
-        sg = sendgrid.SendGridAPIClient(apikey=SENDGRID_API_KEY)
+    email = generate_email(type, email_args, user)
+
+    mail = Mail(
+        from_email=Email(SENDGRID_DEFAULT_FROM),
+        to_email=Email(to),
+        subject=email['info']['subject'],
+    )
+    mail.add_content(Content('text/plain', email['text']))
+    mail.add_content(Content('text/html', email['html']))
+
+    mail.___type = type
+    mail.___to = to
 
-        mail = Mail(
-            from_email=Email(SENDGRID_DEFAULT_FROM),
-            to_email=Email(to),
-            subject=email['info']['subject'],
-        )
-        mail.add_content(Content('text/plain', email['text']))
-        mail.add_content(Content('text/html', email['html']))
+    return mail
 
+
+def sendgrid_send(mail):
+    try:
+        sg = sendgrid.SendGridAPIClient(apikey=SENDGRID_API_KEY)
         res = sg.client.mail.send.post(request_body=mail.get())
-        print('Just sent an email to %s of type %s, response code: %s' % (to, type, res.status_code))
+        print('Just sent an email to %s of type %s, response code: %s' % (mail.___to, mail.___type, res.status_code))
     except HTTPError as e:
-        print('An HTTP error occured while sending an email to %s - %s: %s' % (to, e.__class__.__name__, e))
+        print('An HTTP error occured while sending an email to %s - %s: %s' % (mail.___to, e.__class__.__name__, e))
         print(e.body)
     except Exception as e:
-        print('An unknown error occured while sending an email to %s - %s: %s' % (to, e.__class__.__name__, e))
+        print('An unknown error occured while sending an email to %s - %s: %s' % (mail.___to, e.__class__.__name__, e))
+
+
+class EmailSender(Thread):
```

Comment:
Is my understanding correct that you went the OOP route so that each request would have its own sender queue, and that due to the threaded nature, you could have had a case where two requests were iterating over the same list that was being mutated on the fly?

Reply:
Yes. Tho there are likely other ways of doing it (like attaching email queue to some sort of request context and then finding the correct flask-lifecycle hook, post-flush to client, to start processing said queue).

Reply:
I'm completely satisfied with this approach, and we really only need to do it in batch send instances, so it doesn't really add much code overhead. Merge when ready 👍